### PR TITLE
Move Router dispatch to uniq ROUTE_CHANGE

### PR DESCRIPTION
**Pull request by m4dz**
This PR clean routing and actionsCreators.

- [x] Make initialize runtime more decoupled and simple
- [x] Clean dispatcher
- [x] Use a single `ROUTE_CHANGE` event on URL updates
- [x] Redirect root URL to first account's INBOX
- ~~Clean ActionsCreators~~

### CODE REVIEW

**Comment by misstick** [comment]
@m4dz : I also made some changes that should conflict here. 
 1. which one should be merged first?
 2. however, we must be careful when rebasing our branches (regressions)

See : https://github.com/cozy/cozy-emails/pull/819 and https://github.com/cozy/cozy-emails/pull/814

**Comment by frankrousseau** [comment]
Too much PR open for the end of the sprint. Please merge or close them on monday.

**Comment by m4dz** [comment]
@misstick rebase ok for me, but there's some tests needed when merging with #819/#814. Can we do this on monday morning in peer-review?

File: client/app/router.coffee [hunk 1]
```diff
@@ -17,22 +22,24 @@ _ = require 'lodash'
...
-        'mailbox/:mailboxID/:messageID/reply-all'   : 'messageReplyAll'
-        'mailbox/:mailboxID/:messageID/*'           : 'messageShow'
-        '/*'                                        : 'messageList'
+        'mailbox/:mailboxID/new':                  'messageNew'
+        'mailbox/:mailboxID/:messageID/edit':      'messageEdit'
+        'mailbox/:mailboxID/:messageID/forward':   'messageForward'
+        'mailbox/:mailboxID/:messageID/reply':     'messageReply'
+        'mailbox/:mailboxID/:messageID/reply-all': 'messageReplyAll'
+        'mailbox/:mailboxID/:messageID(?:query)':  'messageShow'
+        # FIXME: set redirect to inbox of first account
```
Comment (misstick) [inline client/app/router.coffee]: @m4dz : shouldnt you remove this comment?

File: client/app/router.coffee [hunk 2]
```diff
@@ -42,52 +49,105 @@ class Router extends Backbone.Router
         # Display application
         _displayApplication()
 
+
+    firstInbox: ->
+        mailboxID = AccountStore.getDefault().get('id')
+        @navigate "mailbox/#{mailboxID}", trigger: true
```
Comment (misstick) [inline client/app/router.coffee]: @m4dz 👍

File: client/app/router.coffee [hunk 3]
```diff
@@ -42,52 +49,105 @@ class Router extends Backbone.Router
...
+
+
     accountNew: ->
-        RouterActionCreator.setAction AccountActions.CREATE
-        AccountActionCreator.selectAccount()
+        # RouterActionCreator.setAction AccountActions.CREATE
+        # AccountActionCreator.selectAccount()
```
Comment (misstick) [inline client/app/router.coffee]: @m4dz : did this route work? If it's OK I think that these comments are useless

File: client/app/router.coffee [hunk 4]
```diff
@@ -42,52 +49,105 @@ class Router extends Backbone.Router
...
+
 
     accountEdit: (accountID, tab) ->
-        RouterActionCreator.setAction AccountActions.EDIT
-        AccountActionCreator.saveEditTab tab
-        AccountActionCreator.selectAccount accountID
+        # RouterActionCreator.setAction AccountActions.EDIT
+        # AccountActionCreator.saveEditTab tab
+        # AccountActionCreator.selectAccount accountID
+        _dispatch {action: AccountActions.EDIT, accountID, tab}
```
Comment (misstick) [inline client/app/router.coffee]: @m4dz : same as above

File: client/app/router.coffee [hunk 5]
```diff
@@ -42,52 +49,105 @@ class Router extends Backbone.Router
...
+
 
     messageList: (mailboxID, query) ->
-        RouterActionCreator.setAction MessageActions.SHOW_ALL
-        LayoutActionCreator.updateMessageList {mailboxID, query}
+        # RouterActionCreator.setAction MessageActions.SHOW_ALL
+        # LayoutActionCreator.updateMessageList {mailboxID, query}
+        _dispatch {action: MessageActions.SHOW_ALL, mailboxID}, query
```
Comment (misstick) [inline client/app/router.coffee]: @m4dz : same as above

File: client/app/router.coffee [hunk 6]
```diff
@@ -42,52 +49,105 @@ class Router extends Backbone.Router
...
+
 
     messageShow: (mailboxID, messageID, query) ->
-        RouterActionCreator.setAction MessageActions.SHOW
-        LayoutActionCreator.updateMessageList {mailboxID, messageID, query}
+        # RouterActionCreator.setAction MessageActions.SHOW
+        # LayoutActionCreator.updateMessageList {mailboxID, messageID, query}
```
Comment (misstick) [inline client/app/router.coffee]: @m4dz : same as above

File: client/app/router.coffee [hunk 7]
```diff
@@ -42,52 +49,105 @@ class Router extends Backbone.Router
...
+        _dispatch {action: MessageActions.SHOW, mailboxID, messageID}, query
+
 
     messageEdit: (mailboxID, messageID) ->
-        RouterActionCreator.setAction MessageActions.EDIT
-        LayoutActionCreator.saveMessage {mailboxID, messageID}
+        # RouterActionCreator.setAction MessageActions.EDIT
+        # LayoutActionCreator.saveMessage {mailboxID, messageID}
```
Comment (misstick) [inline client/app/router.coffee]: @m4dz : same as above

File: client/app/router.coffee [hunk 8]
```diff
@@ -42,52 +49,105 @@ class Router extends Backbone.Router
...
+        _dispatch {action: MessageActions.EDIT, mailboxID, messageID}
+
 
     messageNew: (mailboxID) ->
-        RouterActionCreator.setAction MessageActions.CREATE
-        LayoutActionCreator.saveMessage {mailboxID}
+        # RouterActionCreator.setAction MessageActions.CREATE
+        # LayoutActionCreator.saveMessage {mailboxID}
```
Comment (misstick) [inline client/app/router.coffee]: @m4dz : same as above

File: client/app/router.coffee [hunk 9]
```diff
@@ -42,52 +49,105 @@ class Router extends Backbone.Router
...
+        _dispatch {action: MessageActions.CREATE, mailboxID}
+
 
     messageForward: (mailboxID, messageID) ->
-        RouterActionCreator.setAction MessageActions.FORWARD
-        LayoutActionCreator.saveMessage {mailboxID, messageID}
+        # RouterActionCreator.setAction MessageActions.FORWARD
+        # LayoutActionCreator.saveMessage {mailboxID, messageID}
```
Comment (misstick) [inline client/app/router.coffee]: @m4dz : same as above

File: client/app/router.coffee [hunk 10]
```diff
@@ -42,52 +49,105 @@ class Router extends Backbone.Router
...
+        _dispatch {action: MessageActions.FORWARD, mailboxID}
+
 
     messageReply: (mailboxID, messageID) ->
-        RouterActionCreator.setAction MessageActions.REPLY
-        LayoutActionCreator.saveMessage {mailboxID, messageID}
+        # RouterActionCreator.setAction MessageActions.REPLY
+        # LayoutActionCreator.saveMessage {mailboxID, messageID}
```
Comment (misstick) [inline client/app/router.coffee]: @m4dz : same as above

File: client/app/router.coffee [hunk 11]
```diff
@@ -42,52 +49,105 @@ class Router extends Backbone.Router
...
+        _dispatch {action: MessageActions.REPLY, mailboxID}
+
 
     messageReplyAll: (mailboxID, messageID) ->
-        RouterActionCreator.setAction MessageActions.REPLY_ALL
-        LayoutActionCreator.saveMessage {mailboxID, messageID}
+        # RouterActionCreator.setAction MessageActions.REPLY_ALL
+        # LayoutActionCreator.saveMessage {mailboxID, messageID}
```
Comment (misstick) [inline client/app/router.coffee]: @m4dz : same as above

File: client/app/stores/search_store.coffee [hunk 12]
```diff
@@ -62,9 +62,10 @@ class SearchStore extends Store
     ###
     __bindHandlers: (handle) ->
 
-        handle ActionTypes.QUERY_PARAMETER_CHANGED, ->
-            _resetSearch()
-            @emit 'change'
+        handle ActionTypes.ROUTE_CHANGE, (value) ->
+            if value.query
```
Comment (misstick) [inline client/app/stores/search_store.coffee]: @m4dz : maybe write : `(params={}) ->` and `{query} = params` instead?

**Comment by misstick** [comment]
@m4dz : don't forget to remove dead code

File: client/app/router.coffee [hunk 13]
```diff
@@ -42,52 +49,105 @@ class Router extends Backbone.Router
         # Display application
         _displayApplication()
 
+
+    firstInbox: ->
```
Comment (misstick) [inline client/app/router.coffee]: @m4dz :: prefer `getDefautlMailbox` instead?

**Comment by misstick** [comment]
@m4dz : don't worry about #819/#814 : I will merge this one and rebase mine after

**Pull request merged**